### ISSUE TITLE
Apache MINA SSHD : Client Session Pooling 

### ISSUE THREAD

**Issue by GeezFORCE** [apache/mina-sshd#380]
### Description

Not sure if this is the right platform to ask, please direct to any other platform if not.

Are there any docs/ samples which illustrates how we can pool a predefined number of Client Sessions for use frequently.

### Motivation

I have an use-case for the SFTP client where I need to regularly connect to an SFTP Server for performing an operation. Currently, I create a dedicated client, a session and sftp client each time when I need to do this. But, this is not scalable as the server imposes a restriction on the number of concurrent connections that can be established.

So I thought of implementing a connection pool with a pre-defined set of client sessions all ready to go, similar to Database connection pooling. The idea is when I need to connect to the server, I pick on of these sessions from the pool, do my operation and return it back to the pool, and re-use it later.

### Alternatives considered

I tried to implement my own pooling establishing a fixed number of client sessions. I implemented it as a HashMap with and identifier and corresponding sessions established in a BlockingQueue. I can connect to the server just after initializing the pool, but for subsequent connections, I am finding it hard to keep the sessions alive even by setting up the [Heart Beat API](https://github.com/apache/mina-sshd/blob/master/docs/client-setup.md#keeping-the-session-alive-while-no-traffic) available for the client session object as follows.

`clientSession.setSessionHeartbeat(SessionHeartbeatController.HeartbeatType.IGNORE, TimeUnit.MILLISECONDS, 500L);`

 I get the following exception while trying to re-use the client session again.

`Channel id=1 not registered because session is being closed: SftpChannelSubsystem[id=1, recipient=-1]-ClientSessionImpl[<user>@/<host>:22][sftp]`

Are there any additional steps that needs to be done before setting up the heartbeat? If so, kindly provide the relevant information.

I have tried to change the type of HeartBeat to `RESERVED` thinking of implementing some custom logic, but I cannot wrap my head around how it can be implemented. I have also tried to increase the various timeout values like `IDLE_TIMEOUT`, `SOCKET_KEEPALIVE` etc (I am not sure if these are even applicable for SFTP), but to no avail.

### Additional context

```
MINA Version : 2.9.2
Dependency : sshd-sftp
```

I want to know if this is possible with Apche MINA, and the best practices to implement it. If this has been already done, please direct me to any samples available or kindly share pointers on how this can be implemented

Thanks in advance

**Comment by tomaswolf** [comment]
Several thoughts on that:

- You always have to deal with sessions getting closed, whether or not there is a pool. You cannot control when a server disconnects you. You can try to delay it with heartbeats, but ultimately a connection may go down all the same for any number of reasons, including network failures outside the control of either client or server. A simple approach is: if you get a closed or closing session from the pool, discard it and create a new one.
- If your server has a hard limit on the number of concurrent sessions, keep in mind that any session you open and keep open will prevent someone else from connecting. (Say, the server has a limit of 50, and you create 5 sessions and keep them open: now only 45 other clients can even connect to that server.)
- You could use a single `SshClient`. That would save some per-connection overhead.
- Why do you want to keep your session open for long at all? If you open a session, do your SFTP directory listing, download anything new, and then close the session, why do you even need a pool? Why would that be a problem? (If you're worried about being the 51st session on the next check and thus not able to connect again: you have to handle this case anyway as it might occur already on your very first connection attempt.)
- Why do you even have multiple concurrent sessions per server if you only want to check SFTP directories at one server? Are these directories for different users, requiring different log-ins? Then I don't see how a pool per server would help. And it wouldn't need a pool per user.
- About sessions being closed: how exactly do you check the remote directory? If you're using an `SftpFileSystem`, see Apache MINA sshd bug [SSHD-1310](https://issues.apache.org/jira/browse/SSHD-1310).
- I don't know your system and the constraints, but are you sure that polling via SFTP is the correct architecture? Did you consider making the server tell you when something new has arrived? It could, for instance, post an event or message on an event bus or message bus, and your client just subscribes to that event or message channel or topic, and goes download via SFTP whatever the event says was new. No need for long-lived SSH session with such an architecture. And I'm pretty sure that the top message bus/event bus systems do have SFTP source connectors that can do what you want.

**Comment by GeezFORCE** [comment]
Thank You Tomas for the comprehensive answer.

I can confirm that my custom pool implementation was not successful due to the bug you mentioned. I was using the `SftpFileSystem`.

We don't have control over the way the SFTP server works, so our only option is to periodically check for new files. The motivation behind me trying to implement pooling was our current implementation had to use multiple concurrent sessions for the same user as a consequence of the design.

After reading your answer, I think the better approach will be to have a single connection/ session. Hence I will try to implement that. 

Thanks again for your answer, feel free to close the issue.
